e2e/consul/consul.go

```diff
@@ -29,199 +37,202 @@ func init() {
+			return false, fmt.Errorf("expected %d placed canaries but found %#v",
+				expected, activeDeploy.TaskGroups["consul_canary_test"])
+		}
 
-	// Start a deployment
-	job, _, err := jobs.Info(jobId, nil)
-	require.Nil(err)
-	job.Meta = map[string]string{"version": "2"}
-	resp, _, err := jobs.Register(job, nil)
-	require.Nil(err)
-	require.NotEmpty(resp.EvalID)
+		return true, nil
+	}, func(err error) {
+		t.Fatalf("error while waiting for deploys: %v", err)
+	})
+
+	allocID := activeDeploy.TaskGroups["consul_canary_test"].PlacedCanaries[0]
+	testutil.WaitForResult(func() (bool, error) {
+		alloc, _, err := nomadClient.Allocations().Info(allocID, nil)
+		if err != nil {
+			return false, err
+		}
+
+		if alloc.DeploymentStatus == nil {
+			return false, fmt.Errorf("canary alloc %s has no deployment status", allocID)
+		}
+		if alloc.DeploymentStatus.Healthy == nil {
+			return false, fmt.Errorf("canary alloc %s has no deployment health: %#v",
+				allocID, alloc.DeploymentStatus)
+		}
+		return *alloc.DeploymentStatus.Healthy, fmt.Errorf("expected healthy canary but found: %#v",
+			alloc.DeploymentStatus)
+	}, func(err error) {
+		t.Fatalf("error waiting for canary to be healthy: %v", err)
+	})
 
-	// Eventually have a canary
-	var deploys []*api.Deployment
-	g.Eventually(func() []*api.Deployment {
-		deploys, _, err = jobs.Deployments(*job.ID, false, nil)
-		require.Nil(err)
-		return deploys
-	}, 2*time.Second, 20*time.Millisecond).Should(HaveLen(2))
-
-	deployments := nomadClient.Deployments()
-	var deploy *api.Deployment
-	g.Eventually(func() []string {
-		deploy, _, err = deployments.Info(deploys[0].ID, nil)
-		require.Nil(err)
-		return deploy.TaskGroups["consul_canary_test"].PlacedCanaries
-	}, 2*time.Second, 20*time.Millisecond).Should(HaveLen(1))
-
-	allocations := nomadClient.Allocations()
-	g.Eventually(func() bool {
-		allocID := deploy.TaskGroups["consul_canary_test"].PlacedCanaries[0]
-		alloc, _, err := allocations.Info(allocID, nil)
-		require.Nil(err)
-		return alloc.DeploymentStatus != nil && alloc.DeploymentStatus.Healthy != nil && *alloc.DeploymentStatus.Healthy
-	}, 3*time.Second, 20*time.Millisecond).Should(BeTrue())
-
-	// Query consul catalog for service
-	catalog := consulClient.Catalog()
 	// Check Consul for canary tags
-	g.Eventually(func() []string {
-		consulServices, _, err := catalog.Service("canarytest", "", nil)
-		require.Nil(err)
-
-		for _, serviceInstance := range consulServices {
-			for _, tag := range serviceInstance.ServiceTags {
-				if tag == "canary" {
-					return serviceInstance.ServiceTags
-				}
+	testutil.WaitForResult(func() (bool, error) {
+		consulServices, _, err := consulClient.Catalog().Service("canarytest", "", nil)
+		if err != nil {
+			return false, err
+		}
+		found := false
+		for _, s := range consulServices {
+			if e2eutil.ConsulTagsEqual(s, []string{"canary", "foo"}) {
+				found = true
+				break
```

Comment:
Could this just be an early return? i.e. `return true, nil` and then have `return false fmt.Errorf...` at the bottom of the function

Reply:
+1, if anything I feel `WaitForResult` breaks Go convention by not checking the error return first